File: 70P/How_Many_Numbers_Are_Smaller.py
# my solution 

nums = [8,1,2,2,3]


def count_smalls(numbers):
    ans = []
    for i in range(len(numbers)):
        count = 0
        for j in range(len(numbers)):
            if numbers[j] < numbers[i]:
                count += 1
        ans.append(count)
        count = 0
    return ans   
    
# optimal solution 
def count_smalls_O(numbers):
    temp = sorted(numbers)
    dictt = {}
    ans = []
    
    for index, value in enumerate(temp):
        if value not in dictt:
            dictt[value] = index
    
    for i in numbers:
        ans.append(dictt[i])
    
    return ans 
    
    
# damn that is optimal solution 
def count_smallers(number):
    count = [0] * 102
    
    for num in nums:
        count[num + 1] += 1
    
    for i in range(1, 102):
        count[i] += count[i-1]
    
    return [count[num] for num in nums]


ex1 = count_smalls(nums)
ex2 = count_smallers(nums)
ex3 =  count_smalls_O(nums) 
print(ex1)
print(ex2)   
print(ex3)

File: 70P/How_Many_Numbers_Are_Smaller.js
// time complexity is n * n therefore O(n^2) not efficient, I will implement the optimal one later 


let nums = [8,1,2,2,3]



let countSmallest =  (numbers) => {
    let ans = [];
    const leng = numbers.length

    for(let i = 0; i < leng; i++){
        let count = 0;
        for(let j = 0; j < leng; j++){
            if (numbers[j] < numbers[i]){
                count++;
            }
        }
        ans.push(count);
        count = 0;
    }
    return ans
}

function smallers(numbers){
    const copy = [...numbers];
    const temp = copy.sort((a, b) => a - b);
    let hashMap = {};
    let ans = [];

    for (let i = 0; i < temp.length; i++){
        if(!(temp[i] in hashMap)){
            hashMap[temp[i]] = i;
        }
    }
    for (let j = 0; j < nums.length; j++){
        ans.push(hashMap[nums[j]]);
    }
    return ans;
}
console.log(countSmallest(nums));
console.log(smallers(nums));
